doc/dev/mds_internals/quiesce.rst

```diff
@@ -0,0 +1,90 @@
+MDS Quiesce Protocol
+====================
+
+The MDS quiesce protocol is a mechanism for "quiescing" (quieting) a subvolume
+in a file system, stopping all write (and most read) I/O.  A subvolume is a
+subtree in the file system marked with the ``ceph.dir.subvolume`` flag (like
+the ``ceph-mgr`` volumes plugin subvolumes).
+
+The purpose of this API is to prevent multiple clients from interleaving reads
+and writes across an eventually consistent snapshot barrier where out-of-band
+communication exists between clients. This communication can lead to clients
+wrongly believing they've reached a checkpoint that is mutually recoverable to
+via a snapshot.
+
+Mechanism
+---------
+
+The MDS quiesces I/O using a new ``quiesce_subvolume`` internal request that
+obtains appropriate locks on the root of a subvolume and then launches a series
+of sub-requests for locking other inodes in the subvolume. The locks obtained
+will force clients to release caps and in-progress client/MDS requests to
+complete.
+
+The sub-requests launched are ``quiesce_subvolume_inode`` internal requests
+which simply lock the inode, if the MDS is authoritative for the inode.
+Generally, these are rdlocks (read locks) on each inode metadata lock but the
+``filelock`` is xlocked (exclusively locked) because its type allows for
+multiple readers and writers. Additionally, a new ``quiescelock`` is exclusively
+locked (more on that next).
+
+The inodes in a subvolume are known because subvolumes are a special kind of
+``SnapRealm``. Before the quiesce protocol was written, the MDS already kept
+track of inodes with issued capabilities in the ``SnapRealm``. It is further
+enhanced to keep a list of all inodes in cache.
+
+Because the ``quiesce_subvolume_inode`` request will xlock the ``filelock`` and
+``quiescelock``, it is necessary to run this operation only on the
+authoritative MDS. It is expected that the glue layer on top of the quiesce
+protocol will execute the same ``quiesce_subvolume`` operation on each MDS
+rank. This allows each rank which may be authoritative for part of the
+subvolume subtree to lock all inodes it is authoritative for in the subvolume
+``SnapRealm``.
```

Comment:
good point - we should add the corresponding developer doc for the quiesce db and link to it from here. Maybe just embed a dead link here for now and we'll populate it later

Reply:
I think the docs build complains if we do that.

Reply:
ok. We could then just create a stub page with a TBD section, but that's not critical.